### PR TITLE
dev docs: Adds instructions on running custom agent builds with python checks

### PR DESCRIPTION
**Pull request by scottopell**
### What does this PR do?
Updates some developer-facing documentation with instructions on how to work with the standard python checks.

### Motivation
Knowledge sharing.

### Additional Notes
It would be great if someone could try this on their system and verify.

Particularly the note about `LD_LIBRARY_PATH` being needed, I've never been able to get python loaded without that step, but it feels odd that there are no other notes about this.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by buraizu** [comment]
Thanks for the PR, created DOCS-4267 to track as an editorial review

**Review by remeh — APPROVED** [review]
lgtm

File: docs/dev/checks/README.md [hunk 1]
```diff
@@ -133,3 +125,69 @@ checks code.
...
+## Running checks with a custom Agent Build
+### Custom Checks
+Note:
+If you want to run a custom check inside your development workspace
+(github.com/DataDog/datadog-agent), you must put `MyCheck.yaml` and
+`MyCheck.py` in the `bin/agent/dist` folder located at the root of the
+datadog-agent repository.
+Please keep in mind that the `invoke agent.build` task will copy the
+contents inside `dev/dist` to `bin/agent/dist` when ran, so you can use
+that path if you need alonger-lived location for your custom checks.
```
[the remark this answers is not in the document]
Comment (scottopell) [reply docs/dev/checks/README.md:138]: ya I just copy-pastad this, but it could benefit from a little rewording. 👍

**Review by djmitche — APPROVED** [review]

**Comment by scottopell** [comment]
Ended up re-writing most of this, pls take another look 🙏  (my first version only addressed the base dependencies, not the actual checks)

File: docs/dev/checks/README.md [hunk 2]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+There are a number of checks that ship with a full build of the Agent, but when
+you're developing against a local build, you will not get any of these python
+checks out of the box, you must install them.
+
+> Note the following instructions install python packages to the python user
+> install directory. This could conflict with existing packages.
+
+The list of checks currently shipping with the agent lives in the
+[integrations-core repo
+here](https://github.com/DataDog/integrations-core/blob/master/requirements-agent-release.txt)
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:159]:
```suggestion
The list of checks currently shipping with the Agent lives in the
[integrations-core repo
here](https://github.com/DataDog/integrations-core/blob/master/requirements-agent-release.txt)
```

File: docs/dev/checks/README.md [hunk 3]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+
+Say you want to test the `openmetrics` check.
+
+1. Clone `integrations-core` locally. (Optionally checkout the git tag
+   corresponding to the version you want to test)
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:164]:
```suggestion
   corresponding to the version you want to test.)
```

File: docs/dev/checks/README.md [hunk 4]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+2. From integration-core directory, install `datadog-checks-base` as a pre-req.
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:165]:
```suggestion
2. From the `integrations-core\` directory, install `datadog-checks-base` as a pre-req.
```
I'm guessing this is `integrations-core`, not `integration-core`?

File: docs/dev/checks/README.md [hunk 5]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+   `pip3.10 install --user './datadog_checks_base[deps]'`
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:166]: Generally better for readability/copy-paste to put these in fenced codeblocks, but I can't do that in code-suggestions cos it breaks the suggestion 😅

File: docs/dev/checks/README.md [hunk 6]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+   checkout:
+   `pip3.10 install --user ./openmetrics`
+
+That's it! Your local build should now have the correct packages to be able to
+run the `openmetrics` check.
+
+#### "What is this `[deps]` thing?"
+The `[deps]` at the end of the package name instructs pip to install the
+"requires" that match the `deps` "extra". Without this, you'll see errors when running the agent
+about missing dependencies.
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:177]:
```suggestion
The `[deps]` at the end of the package name instructs pip to install the
requirements that match the `deps` "extra". Without this, you'll see errors 
about missing dependencies when running the Agent.
```

File: docs/dev/checks/README.md [hunk 7]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+
+View all the possible 'extras' with:
+```
+pip3.10 install --no-warn-script-location --user --dry-run --ignore-installed -qqq --report - datadog-checks-base | jq '.install[] | select(.metadata.name=="datadog-checks-base") | .metadata.provides_extra'
+```
+
+View all the "requires" (which shows which "extra" will trigger that
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:184]:
```suggestion
View all the requirements (which shows which "extra" will trigger that
```

File: docs/dev/checks/README.md [hunk 8]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+dependency to be installed) with:
+```
+pip3.10 install --no-warn-script-location --user --dry-run --ignore-installed -qqq --report - datadog-checks-base | jq '.install[] | select(.metadata.name=="datadog-checks-base") | .metadata.requires_dist'
+```
+
+#### "Which python is my agent build using?"
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:190]:
```suggestion
#### "Which Python binary is my Agent build using?"
```

File: docs/dev/checks/README.md [hunk 9]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+Assuming you're not doing a full omnibus build, you won't have an embedded
+python in your local build, so your local build will need to choose a python
+from your system to use.
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:193]:
```suggestion
Assuming you're not doing a full omnibus build, you won't have an embedded
Python binary in your local build, so your local build will need to choose a Python
binary from your system to use.
```

File: docs/dev/checks/README.md [hunk 10]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+
+By default the agent will choose the first python binary (`python` or `python3`
+depending on your configuration) from the `PATH` that is used to run the agent.
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:196]:
```suggestion
By default, the Agent chooses the first Python binary (`python` or `python3`
depending on your configuration) from the `PATH` that is used to run the Agent.
```

File: docs/dev/checks/README.md [hunk 11]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+
+### "Could not initialize Python"
+Out of the box, after an `inv agent.build`, you may see the following error on
+linux when trying to run the resulting `agent` binary:
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:200]:
```suggestion
Linux when trying to run the resulting `agent` binary:
```

File: docs/dev/checks/README.md [hunk 12]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+
+`Could not initialize Python: could not load runtime python for version 3: Unable to open three library: libdatadog-agent-three.so: cannot open shared object file: No such file or directory`
+
+To solve on linux, use the loader env var `LD_LIBRARY_PATH`.
+ie, `LD_LIBRARY_PATH=$PWD/dev/lib ./bin/agent/agent run`
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:205]:
```suggestion
To solve on Linux, use the loader env var `LD_LIBRARY_PATH`.
For example, `LD_LIBRARY_PATH=$PWD/dev/lib ./bin/agent/agent run`
```

File: docs/dev/checks/README.md [hunk 13]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+
+Why is this needed? This is due to a combination of the way `rtloader` works
+and how library loading works on linux.
```
Comment (hestonhoffman) [inline docs/dev/checks/README.md:208]:
```suggestion
and how library loading works on Linux.
```

**Review by remeh — APPROVED** [review]

File: docs/dev/checks/README.md (hunk 10 above)
Comment (sgnn7) [inline docs/dev/checks/README.md:196]: It's a bit more subtle than that. If I remember right, we load python shared lib via rtloader that then loads the default python based on some env vars I think but that may be overkill to explain in this doc.

File: docs/dev/checks/README.md [hunk 14]
```diff
@@ -133,3 +125,91 @@ checks code.
...
+
+The very simplified summary is that `libdatadog-agent-rtloader.so` attempts to load
+`libdatadog-agent-three.so` via `dlopen`, however `libdatadog-agent-rtloader`
+does not have its `RUNPATH` set correctly in local builds, so `dlopen` is unable to find
+`libdatadog-agent-three.so`. Using `LD_LIBRARY_PATH` instructs `dlopen` where to
+search for libraries, so using it sidesteps this issue.
+
```
Comment (sgnn7) [inline docs/dev/checks/README.md:215]: We could add some info on how to use `virtualenv` in the future to not pollute system python but this is great stuff as-is!

File: docs/dev/checks/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (scottopell) [reply docs/dev/checks/README.md:166]: Looks like you're right https://bugs.python.org/issue22295
Updated

File: docs/dev/checks/README.md (hunk 10 above)
Comment (scottopell) [reply docs/dev/checks/README.md:196]: I'll include a brief note about this, it seems the code checks for `PythonHome` and uses that if its set, but I didn't get around to testing this part.

File: docs/dev/checks/README.md (hunk 14 above)
Comment (scottopell) [reply docs/dev/checks/README.md:215]: Yes I'm sure `virtualenv` would be cleaner here, but hopefully this serves as a jumping off point for some future person to write up how to use `virtualenv` correctly here

**Pull request merged**